src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+/// [`RawDocumentRef::get_object_id`](RawDocumentRef::get_object_id) or [`RawDocumentRef::
+/// get_str`](RawDocumentRef::get_str). Note that accessing elements is an O(N) operation, as it
+/// requires iterating through the document from the beginning to find the requested key.
+///
+/// ```
+/// # use bson::raw::{RawDocument, RawError};
+/// let doc = RawDocument::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+/// assert_eq!(doc.get_str("hi")?, Some("y'all"));
+/// # Ok::<(), RawError>(())
+/// ```
+#[derive(Debug)]
+pub struct RawDocumentRef {
+    data: [u8],
+}
+
+impl RawDocumentRef {
+    /// Constructs a new RawDocumentRef, validating _only_ the
+    /// following invariants:
+    ///   * `data` is at least five bytes long (the minimum for a valid BSON document)
+    ///   * the initial four bytes of `data` accurately represent the length of the bytes as
+    ///     required by the BSON spec.
+    ///   * the last byte of `data` is a 0
+    ///
+    /// Note that the internal structure of the bytes representing the
+    /// BSON elements is _not_ validated at all by this method. If the
+    /// bytes do not conform to the BSON spec, then method calls on
+    /// the RawDocument will return RawErrors where appropriate.
+    ///
+    /// ```
+    /// # use bson::raw::{RawDocument, RawError};
+    /// let doc = RawDocumentRef::new(b"\x05\0\0\0\0")?;
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn new<D: AsRef<[u8]> + ?Sized>(data: &D) -> RawResult<&RawDocumentRef> {
+        let data = data.as_ref();
+
+        if data.len() < 5 {
+            return Err(RawError::MalformedValue("document too short".into()));
+        }
+
+        let length = i32_from_slice(&data[..4]);
+
+        if data.len() as i32 != length {
+            return Err(RawError::MalformedValue("document length incorrect".into()));
+        }
+
+        if data[data.len() - 1] != 0 {
+            return Err(RawError::MalformedValue(
+                "document not null-terminated".into(),
+            ));
+        }
+
+        Ok(RawDocumentRef::new_unchecked(data))
+    }
+
+    /// Creates a new Doc referencing the provided data slice.
+    fn new_unchecked<D: AsRef<[u8]> + ?Sized>(data: &D) -> &RawDocumentRef {
+        // SAFETY:
+        //
+        // Dereferencing a raw pointer requires unsafe due to the potential that the pointer is
+        // null, dangling, or misaligned. We know the pointer is not null or dangling due to the
+        // fact that it's created by a safe reference. Converting &[u8] to *const [u8] will be
+        // properly aligned due to them being references to the same type, and converting *const
+        // [u8] to *const RawDocumentRef is aligned due to the fact that the only field in a
+        // RawDocumentRef is a [u8], meaning the structs are represented identically at the byte
+        // level.
+        unsafe { &*(data.as_ref() as *const [u8] as *const RawDocumentRef) }
```

Comment:
Would it not be easier / avoid `unsafe` if we just added a lifetime annotation to `RawDocumentRef` and stored an `&'a [u8]` instead of just a `[u8]`? I'm not sure I understand why this is preferable.

Reply:
This makes it possible to implement Deref.  The unsafe pattern is taken straight from the way Path is implemented.

Before this, all the methods on RawDocumentRef were duplicated on RawDocument with .as_ref().method() implementations.

Reply:
The other benefit is that having a Buffer type and an unsized reference type is a familiar pattern from Vec<T> / [T], String / str, PathBuf / Path , OsString / OsStr, etc.

Reply:
Finally, storing an &'a [u8] forces this type to only be used as a reference type.  An usized RawDocumentRef type can also be stored in a Box / Rc / Arc / other smart pointer, and be passed around as an owned type, saving an extra pointer indirection over boxing a RawDocument.

Reply:
>Finally, storing an &'a [u8] forces this type to only be used as a reference type. An usized RawDocumentRef type can also be stored in a Box / Rc / Arc / other smart pointer, and be passed around as an owned type, saving an extra pointer indirection over boxing a RawDocument.

How could we create e.g. a `Box<RawDocumentRef>`? And what's the use case for it?

The first two reasons seem valid, though I'd personally prefer to avoid using `unsafe` / stick with simpler reference based code and just have to deal with the boilerplate of re-implementing all those methods.

Any thoughts on this? @saghm @isabelatkinson

Reply:
I think a compelling example use case would be to be able to do something like `Cow<RawDocumentRef>` in the same way that one could do `Cow<str>`. By giving `RawDocumentRef` the same semantics as existing owned/borrowed pairs like `Vec<T>`/`[T]` and, `String`/`str`, we allow users to re-use the same patterns that they're used to. I'm not thrilled with the idea of having to include unsafe code in the library, but a single line of unsafe limited to very straightforward casting seems like it wouldn't be the end of the world.

Reply:
> How could we create e.g. a Box<RawDocumentRef>? And what's the use case for it?

I think, given that `Deref<Target=RawDocumentRef>` is implemented, you can get a `Box<RawDocumentRef>` by using `Box::new(*raw_document)`.  The usecase, particularly for Rc/Arc, is to get multiple ownership semantics without introducing an extra pointer indirection from inside RawDocument.  The usecase would be for a minor performance and size improvement when passing between tasks/threads in an async/multithreaded context, or storing in data structures where you want need multiple ownership.

I also personally find it harder to read struct lifetimes than simple reference lifetimes.

It's definitely a maintenance question for y'all to answer primarily.  One line of unsafe to audit vs maintaining duplicate APIs.

Reply:
`Box::new(*raw_document)` doesn't work since `RawDocumentRef` is dynamically sized. That being said, `Cow` does seem to be a pretty good reason to model this type that way. 

If we're going to go this route, we'll need to add a manual `From` implementation for our type, [like they do for other similar reference / slice types](https://doc.rust-lang.org/std/borrow/enum.Cow.html#trait-implementations). Otherwise, creating a `Cow` for our document/reference type can be a little thorny.